File: e2e/java/src/test/java/dev/kreuzberg/e2e/RedirectTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import dev.kreuzberg.kreuzcrawl.Kreuzcrawl;

/** E2e tests for category: redirect. */
class RedirectTest {
    @Test
    void testRedirect301Permanent() throws Exception {
        // Follows 301 permanent redirect and returns final page content
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirect302Found() throws Exception {
        // Follows 302 Found redirect correctly
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirect303SeeOther() throws Exception {
        // Follows 303 See Other redirect (method changes to GET)
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirect307Temporary() throws Exception {
        // Follows 307 Temporary Redirect (preserves method)
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirect308Permanent() throws Exception {
        // Follows 308 Permanent Redirect (preserves method)
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirectChain() throws Exception {
        // Follows a chain of redirects (301 -> 302 -> 200)
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirectCrossDomain() throws Exception {
        // Reports cross-domain redirect target without following to external domain
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirectLoop() throws Exception {
        // Detects redirect loop (A -> B -> A) and returns error
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'is_error' not available on result type
    }

    @Test
    void testRedirectMaxExceeded() throws Exception {
        // Aborts when redirect count exceeds max_redirects limit
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'is_error' not available on result type
    }

    @Test
    void testRedirectMetaRefresh() throws Exception {
        // Follows HTML meta-refresh redirect to target page
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirectRefreshHeader() throws Exception {
        // Handles HTTP Refresh header redirect
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
    }

    @Test
    void testRedirectTo404() throws Exception {
        // Redirect target returns 404 Not Found
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'final_url' not available on result type
        // skipped: field 'redirect_count' not available on result type
        // skipped: field 'is_error' not available on result type
    }

}

File: e2e/go/redirect_test.go
// E2e tests for category: redirect
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_Redirect301Permanent(t *testing.T) {
	// Follows 301 permanent redirect and returns final page content
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_Redirect302Found(t *testing.T) {
	// Follows 302 Found redirect correctly
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_Redirect303SeeOther(t *testing.T) {
	// Follows 303 See Other redirect (method changes to GET)
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_Redirect307Temporary(t *testing.T) {
	// Follows 307 Temporary Redirect (preserves method)
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_Redirect308Permanent(t *testing.T) {
	// Follows 308 Permanent Redirect (preserves method)
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_RedirectChain(t *testing.T) {
	// Follows a chain of redirects (301 -> 302 -> 200)
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_RedirectCrossDomain(t *testing.T) {
	// Reports cross-domain redirect target without following to external domain
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_RedirectLoop(t *testing.T) {
	// Detects redirect loop (A -> B -> A) and returns error
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'is_error' not available on result type
}

func Test_RedirectMaxExceeded(t *testing.T) {
	// Aborts when redirect count exceeds max_redirects limit
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'is_error' not available on result type
}

func Test_RedirectMetaRefresh(t *testing.T) {
	// Follows HTML meta-refresh redirect to target page
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_RedirectRefreshHeader(t *testing.T) {
	// Handles HTTP Refresh header redirect
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
}

func Test_RedirectTo404(t *testing.T) {
	// Redirect target returns 404 Not Found
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'final_url' not available on result type
	// skipped: field 'redirect_count' not available on result type
	// skipped: field 'is_error' not available on result type
}
